### ISSUE TITLE
Nuxt universal + prod - SSR/service plugins broken after page refresh

### ISSUE THREAD

**Issue by nakedgun** [feathersjs-ecosystem/feathers-vuex#470]
### Steps to reproduce

Setup a very basic nuxt + feathers-vuex example as per the [docs](https://vuex.feathersjs.com/nuxt.html). Nuxt mode should be universal. No auth code necessary, but add at least one service using the makeServicePlugin, and have it wired to a working feathers API with some dummy data. 

Setup a basic page to test SSR - here, we're going to call a feathers-vuex service method in fetch():

```
// pages/index.vue
<template>
 <div>
   <pre>{{ $store.state.conversations }}</pre>
 </div>
</template>
<script>
import { models } from 'feathers-vuex';
export default {
  data: () => ({}),
  async fetch({ store, params }) {
    console.log('start fetch() | store.state.conversations = ', store.state.conversations);
    const { conversation } = models.api;
    await conversation.get(1);
    console.log('end fetch() | store.state.conversations = ', store.state.conversations);
  }
};
</script>
```

1. Start a dev server (e.g. yarn dev) - visit the page & refresh.

![fw-ok](https://user-images.githubusercontent.com/5262234/80540388-0dd1b600-895e-11ea-8260-98ada70a8f6b.gif)

^ All good, store is clean and then populated on each request. SSR works.

1. Now start a production server, e.g. `yarn build && yarn start`
2. Visit the test page - trying refreshing

![fw-bug](https://user-images.githubusercontent.com/5262234/80540441-2cd04800-895e-11ea-902f-829aeaae097e.gif)

### Expected behavior
SSR works on every request.

### Actual behavior
SSR breaks after first request (store state is only being correctly populated on the first request).

### Debugging

If you add a [mutation logger plugin](https://vuex.vuejs.org/guide/plugins.html) to vuex, you'll see that in dev mode we get the correct cycle of state being cleared and populated on each page request:

```

**First request**

begin fetch() | store.state.conversations =  {
  ids: [],
  keyedById: {},
  copiesById: {},
  tempsById: {},
  tempsByNewId: {},
  pagination: { defaultLimit: null, defaultSkip: null },
  isFindPending: false,
  isGetPending: false,
  isCreatePending: false,
  isUpdatePending: false,
  isPatchPending: false,
  isRemovePending: false,
  errorOnFind: null,
  errorOnGet: null,
  errorOnCreate: null,
  errorOnUpdate: null,
  errorOnPatch: null,
  errorOnRemove: null,
  modelName: 'conversation',
  namespace: 'conversations',
  servicePath: 'conversations',
  autoRemove: false,
  addOnUpsert: false,
  enableEvents: false,
  idField: 'id',
  tempIdField: '__id',
  debug: false,
  keepCopiesInStore: false,
  nameStyle: 'short',
  paramsForServer: [],
  preferUpdate: false,
  replaceItems: false,
  serverAlias: 'api',
  skipRequestIfExists: false,
  whitelist: []
}
conversations/setPending get
conversations/addItem Conversation {
  id: 7,
  title: null,
  creatorId: 5,
  recipientId: 1,
  postId: 10011,
  createdAt: '2020-04-23T22:54:26.062Z',
  updatedAt: '2020-04-23T22:54:26.062Z'
}
conversations/addItem Conversation {
  id: 7,
  title: null,
  creatorId: 5,
  recipientId: 1,
  postId: 10011,
  createdAt: '2020-04-23T22:54:26.062Z',
  updatedAt: '2020-04-23T22:54:26.062Z'
}
conversations/unsetPending get
end fetch() | store.state.conversations {
  ids: [ 7 ],
  keyedById: {
    '7': Conversation {
      id: 7,
      title: null,
      creatorId: 5,
      recipientId: 1,
      postId: 10011,
      createdAt: '2020-04-23T22:54:26.062Z',
      updatedAt: '2020-04-23T22:54:26.062Z'
    }
  },
  copiesById: {},
  tempsById: {},
  tempsByNewId: {},
  pagination: { defaultLimit: null, defaultSkip: null },
  isFindPending: false,
  isGetPending: false,
  isCreatePending: false,
  isUpdatePending: false,
  isPatchPending: false,
  isRemovePending: false,
  errorOnFind: null,
  errorOnGet: null,
  errorOnCreate: null,
  errorOnUpdate: null,
  errorOnPatch: null,
  errorOnRemove: null,
  modelName: 'conversation',
  namespace: 'conversations',
  servicePath: 'conversations',
  autoRemove: false,
  addOnUpsert: false,
  enableEvents: false,
  idField: 'id',
  tempIdField: '__id',
  debug: false,
  keepCopiesInStore: false,
  nameStyle: 'short',
  paramsForServer: [],
  preferUpdate: false,
  replaceItems: false,
  serverAlias: 'api',
  skipRequestIfExists: false,
  whitelist: []
}

***Page Reload***

begin fetch() | store.state.conversations =  {
  ids: [],
  keyedById: {},
  copiesById: {},
  tempsById: {},
  tempsByNewId: {},
  pagination: { defaultLimit: null, defaultSkip: null },
  isFindPending: false,
  isGetPending: false,
  isCreatePending: false,
  isUpdatePending: false,
  isPatchPending: false,
  isRemovePending: false,
  errorOnFind: null,
  errorOnGet: null,
  errorOnCreate: null,
  errorOnUpdate: null,
  errorOnPatch: null,
  errorOnRemove: null,
  modelName: 'conversation',
  namespace: 'conversations',
  servicePath: 'conversations',
  autoRemove: false,
  addOnUpsert: false,
  enableEvents: false,
  idField: 'id',
  tempIdField: '__id',
  debug: false,
  keepCopiesInStore: false,
  nameStyle: 'short',
  paramsForServer: [],
  preferUpdate: false,
  replaceItems: false,
  serverAlias: 'api',
  skipRequestIfExists: false,
  whitelist: []
}
conversations/setPending get
conversations/addItem Conversation {
  id: 7,
  title: null,
  creatorId: 5,
  recipientId: 1,
  postId: 10011,
  createdAt: '2020-04-23T22:54:26.062Z',
  updatedAt: '2020-04-23T22:54:26.062Z'
}
conversations/addItem Conversation {
  id: 7,
  title: null,
  creatorId: 5,
  recipientId: 1,
  postId: 10011,
  createdAt: '2020-04-23T22:54:26.062Z',
  updatedAt: '2020-04-23T22:54:26.062Z'
}
conversations/unsetPending get
end fetch() | store.state.conversations {
  ids: [ 7 ],
  keyedById: {
    '7': Conversation {
      id: 7,
      title: null,
      creatorId: 5,
      recipientId: 1,
      postId: 10011,
      createdAt: '2020-04-23T22:54:26.062Z',
      updatedAt: '2020-04-23T22:54:26.062Z'
    }
  },
  copiesById: {},
  tempsById: {},
  tempsByNewId: {},
  pagination: { defaultLimit: null, defaultSkip: null },
  isFindPending: false,
  isGetPending: false,
  isCreatePending: false,
  isUpdatePending: false,
  isPatchPending: false,
  isRemovePending: false,
  errorOnFind: null,
  errorOnGet: null,
  errorOnCreate: null,
  errorOnUpdate: null,
  errorOnPatch: null,
  errorOnRemove: null,
  modelName: 'conversation',
  namespace: 'conversations',
  servicePath: 'conversations',
  autoRemove: false,
  addOnUpsert: false,
  enableEvents: false,
  idField: 'id',
  tempIdField: '__id',
  debug: false,
  keepCopiesInStore: false,
  nameStyle: 'short',
  paramsForServer: [],
  preferUpdate: false,
  replaceItems: false,
  serverAlias: 'api',
  skipRequestIfExists: false,
  whitelist: []
}
```

For prod -- only the first page request cycle is healthy:

```

**First request**

begin fetch() | store.state.conversations =  {
  ids: [],
  keyedById: {},
  copiesById: {},
  tempsById: {},
  tempsByNewId: {},
  pagination: { defaultLimit: null, defaultSkip: null },
  isFindPending: false,
  isGetPending: false,
  isCreatePending: false,
  isUpdatePending: false,
  isPatchPending: false,
  isRemovePending: false,
  errorOnFind: null,
  errorOnGet: null,
  errorOnCreate: null,
  errorOnUpdate: null,
  errorOnPatch: null,
  errorOnRemove: null,
  modelName: 'conversation',
  namespace: 'conversations',
  servicePath: 'conversations',
  autoRemove: false,
  addOnUpsert: false,
  enableEvents: false,
  idField: 'id',
  tempIdField: '__id',
  debug: false,
  keepCopiesInStore: false,
  nameStyle: 'short',
  paramsForServer: [],
  preferUpdate: false,
  replaceItems: false,
  serverAlias: 'api',
  skipRequestIfExists: false,
  whitelist: []
}
conversations/setPending get
conversations/addItem o {
  id: 7,
  title: null,
  creatorId: 5,
  recipientId: 1,
  postId: 10011,
  createdAt: '2020-04-23T22:54:26.062Z',
  updatedAt: '2020-04-23T22:54:26.062Z'
}
conversations/addItem o {
  id: 7,
  title: null,
  creatorId: 5,
  recipientId: 1,
  postId: 10011,
  createdAt: '2020-04-23T22:54:26.062Z',
  updatedAt: '2020-04-23T22:54:26.062Z'
}
conversations/unsetPending get
end fetch() | store.state.conversations {
  ids: [ 7 ],
  keyedById: {
    '7': o {
      id: 7,
      title: null,
      creatorId: 5,
      recipientId: 1,
      postId: 10011,
      createdAt: '2020-04-23T22:54:26.062Z',
      updatedAt: '2020-04-23T22:54:26.062Z'
    }
  },
  copiesById: {},
  tempsById: {},
  tempsByNewId: {},
  pagination: { defaultLimit: null, defaultSkip: null },
  isFindPending: false,
  isGetPending: false,
  isCreatePending: false,
  isUpdatePending: false,
  isPatchPending: false,
  isRemovePending: false,
  errorOnFind: null,
  errorOnGet: null,
  errorOnCreate: null,
  errorOnUpdate: null,
  errorOnPatch: null,
  errorOnRemove: null,
  modelName: 'conversation',
  namespace: 'conversations',
  servicePath: 'conversations',
  autoRemove: false,
  addOnUpsert: false,
  enableEvents: false,
  idField: 'id',
  tempIdField: '__id',
  debug: false,
  keepCopiesInStore: false,
  nameStyle: 'short',
  paramsForServer: [],
  preferUpdate: false,
  replaceItems: false,
  serverAlias: 'api',
  skipRequestIfExists: false,
  whitelist: []
}

**Page Reload**

begin fetch() | store.state.conversations =  {
  ids: [],
  keyedById: {},
  copiesById: {},
  tempsById: {},
  tempsByNewId: {},
  pagination: { defaultLimit: null, defaultSkip: null },
  isFindPending: false,
  isGetPending: false,
  isCreatePending: false,
  isUpdatePending: false,
  isPatchPending: false,
  isRemovePending: false,
  errorOnFind: null,
  errorOnGet: null,
  errorOnCreate: null,
  errorOnUpdate: null,
  errorOnPatch: null,
  errorOnRemove: null,
  modelName: 'conversation',
  namespace: 'conversations',
  servicePath: 'conversations',
  autoRemove: false,
  addOnUpsert: false,
  enableEvents: false,
  idField: 'id',
  tempIdField: '__id',
  debug: false,
  keepCopiesInStore: false,
  nameStyle: 'short',
  paramsForServer: [],
  preferUpdate: false,
  replaceItems: false,
  serverAlias: 'api',
  skipRequestIfExists: false,
  whitelist: []
}
conversations/setPending get
conversations/mergeInstance {
  id: 7,
  title: null,
  creatorId: 5,
  recipientId: 1,
  postId: 10011,
  createdAt: '2020-04-23T22:54:26.062Z',
  updatedAt: '2020-04-23T22:54:26.062Z'
}
conversations/updateItem o {
  id: 7,
  title: null,
  creatorId: 5,
  recipientId: 1,
  postId: 10011,
  createdAt: '2020-04-23T22:54:26.062Z',
  updatedAt: '2020-04-23T22:54:26.062Z'
}
conversations/unsetPending get
end fetch() | store.state.conversations {
  ids: [],
  keyedById: {},
  copiesById: {},
  tempsById: {},
  tempsByNewId: {},
  pagination: { defaultLimit: null, defaultSkip: null },
  isFindPending: false,
  isGetPending: false,
  isCreatePending: false,
  isUpdatePending: false,
  isPatchPending: false,
  isRemovePending: false,
  errorOnFind: null,
  errorOnGet: null,
  errorOnCreate: null,
  errorOnUpdate: null,
  errorOnPatch: null,
  errorOnRemove: null,
  modelName: 'conversation',
  namespace: 'conversations',
  servicePath: 'conversations',
  autoRemove: false,
  addOnUpsert: false,
  enableEvents: false,
  idField: 'id',
  tempIdField: '__id',
  debug: false,
  keepCopiesInStore: false,
  nameStyle: 'short',
  paramsForServer: [],
  preferUpdate: false,
  replaceItems: false,
  serverAlias: 'api',
  skipRequestIfExists: false,
  whitelist: []
}
```

You'll notice in prod after the page refresh the store state is messed up. We get empty state both before and after fetch() completes and the service module attempts to update records instead of adding them.

I haven't dug much deeper than this, but I noted the globalModels object is not clean on the second request (in prod - it's fine in dev..?!). It has a leftover reference to the conversations service:

**Overwriting Model: models[api][conversation]** (global-models.js)

Annnnd, within service-module-actions.js, it's using this dirty state. Within the page component, we're getting something different. 

Any help would be appreciated! And thanks for the awesome plugin, it's been a big time saver.

### System configuration

Tell us about the applicable parts of your setup.

**Module versions** (especially the part that's not working):

**NodeJS version**:

12.16.1
```
  "dependencies": {
    "@feathersjs/rest-client": "^4.3.11",
    "@nuxtjs/axios": "^5.6.0",
    "@nuxtjs/dotenv": "^1.4.1",
    "@vue/composition-api": "^0.5.0",
    "consola": "^2.10.1",
    "cross-env": "^5.2.0",
    "express": "^4.16.4",
    "feathers-client": "^2.4.0",
    "feathers-hooks-common": "^4.20.7",
    "nuxt-client-init-module": "^0.1.8",
    "feathers-vuex": "^3.9.1",
    "lodash": "^4.17.15",
    "nuxt": "^2.11.0",
  },
```
**Operating System**:
OS X 10.14.6
**Browser Version**:
Firefox/Chrome

**Comment by nakedgun** [comment]
What what I can tell, though I can't find any documentation on this, universal dev and prod builds for Nuxt behave very differently wrt to plugins. In prod, plugins/modules will be loaded _once_ on the server, for the entire node process. In dev, they are reloaded on every request.

The issue (I think) is that there's a couple of global references which are being kept in memory on the server and being re-used in subsequent requests, specifically:

-- globalModels 
-- globalClients

So, whilst vuex module state is reset, _store.state_ in the BaseModel constructor uses state from the previous request. That, and the fact we have left over refs in general, causes the weirdness above (and pretty much breaks everything from this point on).

Anyway, that's all I got. I'm still a bit wtf about the dev/prod difference - hoping someone who knows the internals of feathers-vuex/nuxt could shed some more light on this!

**Comment by nakedgun** [comment]
Alright, it seems like the problem is with makeServicePlugin() in **make-service-plugin.ts**. With production SSR builds, the model will already exist and contain state from the previous request. This state is _dirty_, whereas the vuex module state is _clean_. This state difference causes a lot of bugs.

I believe the fix is relatively safe:

```
// (2b^) Monkey patch the Model(s) and add to globalModels
      assignIfNotPresent(Model, {
        store,
        namespace: options.namespace,
        servicePath,
        instanceDefaults,
        setupInstance,
        preferUpdate
      })
```
Replace with:
```
 // (2b^) Monkey patch the Model(s) and add to globalModels
      assignIfNotPresent(Model, {
        namespace: options.namespace,
        servicePath,
        instanceDefaults,
        setupInstance,
        preferUpdate
      })
      // always assign default state to the model, to prevent state reuse for SSR 
      Object.assign(Model, {
        store
      })
```

**Comment by marshallswain** [comment]
@nakedgun this seems like a simple solution that shouldn't affect other apps, since this code should generally only run once per initialization.  On the client size, the initialization only happens once, so existing apps shouldn't be affected.  Have you checked that this fixes your SSR bug?

**Comment by nakedgun** [comment]
@marshallswain Pretty much. For SSR, that code gets executed on each request. SPA - typically only once, unless someone is re-initializing (I don't see the use case for this on the client though?). Irrespective of the mode, it's just important the Model state and vuex state are in sync. Right now, store.registerModule (in step 1^) is omitting the [preserveState](https://vuex.vuejs.org/guide/modules.html#preserving-state) param, so it's defaulting to clearing state - whereas in 2b^, state is preserved if it exists on the Model. The state difference is the source of the bug.

In theory, preserving state could be useful in some SSR scenarios, so this could be an option for the service. But unless that's requested, a safer default option IMO is to clean state, since it prevents memory leaks for SSR.

And yes - it fixes my app in SPA/SSR (I use both for mobile/web builds). I've also tested memory leaks when spamming SSR routes and it seems fine!

**Comment by marshallswain** [comment]
This sounds good to me. Let’s make it the default and if somebody complains we will give them an option to switch back.  Care to make a PR?

**Comment by nakedgun** [comment]
Sounds good! I'll try and get a PR in this week.

**Comment by marshallswain** [comment]
Fixed in `feathers-vuex@3.10.2`.

**Comment by phatj** [comment]
Thank you both for your work on this issue.

This change does introduce a minor annoyance for Nuxt SSR whereby it constantly complains that the state is overridden (while in dev mode, mind you; in production, the warning is suppressed via [vuex](https://github.com/vuejs/vuex/blob/b2df4db582418ea7a9fbc8c7e46d6486f335d72b/src/store.js#L351)).

The specific warning is caused by setting `preserveState` to false as the merged PR does.  From what I can gather:
* Nuxt SSR will hydrate the state into the store automatically (this is taken from `window.__NUXT__.state`)
* feathers-vuex then replaces the state on module registration (because of the new `preserveState` setting

Regarding this part:
> Right now, store.registerModule (in step 1^) is omitting the preserveState param, so it's defaulting to clearing state

Within, Nuxt, `preserveState` is actually set conditionally in the [template/index.js](https://github.com/nuxt/nuxt.js/blob/dev/packages/vue-app/template/index.js#L79) file.  It just overwrites the function and defaults it to the boolean `process.client` (`false` on server, `true` on client); however, it does allow you to override that should you need to.

As for what we can do in feathers-vuex, I suggest we expose the options for `registerModule` on the `makeServicePlugin` interface at the very least so that other non-Nuxt SSR applications can enable/disable the setting as they see fit, and Nuxt itself can handle the setting naturally.

Let me know if any help is needed :)

**Comment by nakedgun** [comment]
Hey @phatj ! Can you post the warning that you are getting for state being overridden? I've never gotten this for a service module -- only the _auth_ module, which is separate to this flow.

As for your suggestion, I agree in principal that it would be nice to make this configurable. However, I did a quick test (by setting `preserveState: true` and  _preserving_ model state in (2b^) so the two settings are in sync) and I get a crash on load in SPA mode:

```
 TypeError: "state is undefined"
    setPending service-module.mutations.js:340
```

I haven't looked into this any further -- there's probably a way to make this all work nicely with options, but it might take a little more work (and testing - the error above might be specific to my app, not sure without more debugging). I'd also see what @marshallswain feels about this. If we do make this configurable it's important that existing SPA/SSR setups are unaffected.

**Comment by phatj** [comment]
Hi @nakedgun, thanks for responding so quickly!  My apologies for the late reply, my notifications were buried, haha.

This is the warning I got:
```
[vuex] state field "users" was overridden by a module with the same name at "users"
```

I've since wrapped the plugin call with my own `registerModule` hack and everything is nice and peachy again.

That's quite an interesting pickle with the SPA mode.  I've attempted to replicate it in my own repo, but haven't had any luck with it yet.  Would it be possible to work out of a public repo so we can replicate our findings?

I dove into [this section](https://github.com/feathersjs-ecosystem/feathers-vuex/blob/master/src/service-module/make-service-plugin.ts#L97) as well as the global model object.

```typescript
      if (BaseModel && !BaseModel.store) {
        Object.assign(BaseModel, {
          store
        })
      }
      // snipped for brevity...
      // As per 1^, don't preserve state on the model either (prevents state pollution in SSR)
      Object.assign(Model, {
        store
      })
```

Investigation updates:
1. Models are stored globally—when running in SSR mode, subsequent (server-side) requests will overwrite the store (and state).  In general, when writing universal code, nothing can be kept in the global space, it must always be constrained within a request context.  See the [Vue SSR guide](https://ssr.vuejs.org/guide/structure.html#avoid-stateful-singletons) for more information.
2. The static property `store` of the `BaseModel` is ensured to exist which means that subsequent models also have this property set as we are dealing with the class itself and not an instance.  In SSR mode, without ` Object.assign(Model, { store })`, the previous request's state would have been stuck in a globally accessible variable.

You can also verify that the global object is overriding by setting the debug option to true on `global-model.js`.  It will then render this in production mode:
```
 ERROR  Overwriting Model: models[api][User].  
```
![image](https://user-images.githubusercontent.com/3116833/83962099-335cb400-a868-11ea-84b5-bbdafeabca87.png)
> This is from an unauthenticated request without the 3.10.2 patch.

Prior to 3.10.2, this also leaks state between requests on the server side which I'm betting is what led to the original issue that kicked off this thread.

I once had a similar global state in another project, and couldn't figure out for the life of me why my accessTokens were always stuck on an old value 😄.

I think we've got ourselves quite the spicy pickle!

**Comment by nakedgun** [comment]
R.e. global models - for sure, they're an anti-pattern in SSR. My guess is feathers-vuex started off as a client side lib and then evolved to support nuxt/SSR later. 

>  I've attempted to replicate it in my own repo, but haven't had any luck with it yet. Would it be possible to work out of a public repo so we can replicate our findings?

Would probably be best! I can't repro with my codebase -- if you have the time, something demonstrating the issues would be great. The auth one is particularly bad. 

> [vuex] state field "users" was overridden by a module with the same name at "users"

Hard to be certain without seeing the code, but my hunch is you have `users` as a service module (`services/user`) and also as a folder in `store`, e.g. `store/users`.

**Comment by phatj** [comment]
> Would probably be best! I can't repro with my codebase -- if you have the time, something demonstrating the issues would be great. The auth one is particularly bad.

I'll take a look into this today.  There are some hoops to jump through as the official documentation for nuxt integration with socket.io also relies on global state (which is problematic in particular for the socket.io transport as authentication is saved on the connection itself, iirc).

I may try with a RESTful transport only and see if the same errors pop up.

> Hard to be certain without seeing the code, but my hunch is you have users as a service module (services/user) and also as a folder in store, e.g. store/users.

Unfortunately, I only have the `users` service instantiated in a project without any other vuex modules under `./store`.

**Comment by phatj** [comment]
Please see the repo pair here, @nakedgun:
- https://github.com/phatj/nuxt-feathers-vuex-example-api
- https://github.com/phatj/nuxt-feathers-vuex-example-client

This setup follows the spirit of the official docs, but encloses most setup/config steps in request contexts.  It's pared down from a couple of my production projects.

Copy the `.env.default` file to `.env` to get it started.

From here, you should see (in the console), the warning in dev mode as well as the error in SSR mode after running the following commands, respectively.
```bash
# dev warning
yarn dev

# SSR error
yarn build && DEBUG=true yarn start
```

I'm fairly certain that the global objects are the culprits.  The only ways I see getting around this are either:
- monkey patch the global model to operate within the request context (which is what I'm probably going to do in the meantime);
- transform the `import { models } from 'feathers-vuex'` syntactic sugar to work like vuex `mapGetters` (they pick up `this` from the context they are called from; quite ingenious and clean) —this is a BC break, unfortunately; or
- just officially ignore SSR and leave it to userland implementations to solve.

Cheers!

**Comment by nakedgun** [comment]
@phatj you're probably right here. I played around with your repos and get the same error -- sadly right now I don't have the cycles to dig much deeper. Perhaps if you could simplify the repro down to the bare minimum and create a separate case someone from the core team might be able to help (I feel like it'll get lost in sea in this case). FYI, I can't reproduce this with my own codebase, which I would say is closer to the docs than yours.

**Comment by phatj** [comment]
Thanks, @nakedgun.

I've forked this codebase and started to get deeper into the mechanics of feathers-vuex.  Unfortunately, it will require an extensive refactor to encapsulate the global variables:
- moving global objects back into context
- shifting the BaseModel to use dependency injection rather than overwriting it's static store variable
- injecting the models into the store (which probably forces us to remove arrow function expressions from service stores)
- nuxtjs tests (not a fun time, haha) for SSR production issues

It's worth noting that outside of SSR, this library functions wonderfully as intended.

The reason you would not be able to replicate these errors in your codebase is most likely because you are using a global context for the feathers client, feathers vuex, the transport mechanism, etc. as the docs recommend.  In this case, no overwriting occurs, but you are left with this very issue (#470) and `3.10.2` attempts to fix.

Theoretically, I believe this will lead to race conditions in prod SSR depending on async call and request timing.  I currently do not recommend running a production SSR instance for all of my projects/clients because of this tricky issue.  Besides, generating static sites and SPAs is the norm from what I can see.

Unless I can figure out a more general solution, I'll open up a new issue that references this one.  Thank you very much for your time and attention, I appreciate it! 👍

**Comment by nakedgun** [comment]
@phatj One last thing I might add - I don't know your setup or constraints, but another approach here is to use a serverless function for SSR instead. Since they are - by design - short lived processes, you don't have to worry about state pollution (or other annoying SSR gotchas) between requests. Once the client has loaded the SSR bundle, it can then talk to your API server, and pretend like none of this SSR nonsense ever existed.

The other main benefit of serverless here is out of box scaling. Vercel supports [SSR for Nuxt](https://github.com/nuxt/now-builder) (though you can roll your own using AWS lambda + serverless framework). I am leaning towards this route instead, since it generally feels safer than trying to make every 3rd party library I want to use fully SSR compliant.
